### PR TITLE
Fix Issue 1787 "The specified Visual is not an ancestor of this Visual" exception

### LINKED ISSUE #1787: "The specified Visual is not an ancestor of this Visual" exception in OxySKElement.OnRender()
### Description
We use a docking manager in our application. With it it's possible to dock out windows with plots and arrange them on another screens.
With the current oxyplot version and OxyPlot.SkiaSharp.Wpf.PlotView, we get an exception after undocking a window with a plot.

Partial stack-trace:
```
System.InvalidOperationException
  HResult=0x80131509
  Message=The specified Visual is not an ancestor of this Visual.
  Source=PresentationCore
  StackTrace:
   at System.Windows.Media.Visual.TrySimpleTransformToAncestor(Visual ancestor, Boolean inverse, GeneralTransform& generalTransform, Matrix& simpleTransform)
   at System.Windows.Media.Visual.TransformToAncestor(Visual ancestor)
   at OxyPlot.SkiaSharp.Wpf.OxySKElement.OnRender(DrawingContext drawingContext)
   at System.Windows.UIElement.Arrange(Rect finalRect)
   at System.Windows.Controls.Grid.ArrangeOverride(Size arrangeSize)
   at System.Windows.FrameworkElement.ArrangeCore(Rect finalRect)
   at System.Windows.UIElement.Arrange(Rect finalRect)
    ...
    [Call Stack Truncated]

  This exception was originally thrown at this call stack:
    System.Windows.Media.Visual.TrySimpleTransformToAncestor(System.Windows.Media.Visual, bool, out System.Windows.Media.GeneralTransform, out System.Windows.Media.Matrix)
    System.Windows.Media.Visual.TransformToAncestor(System.Windows.Media.Visual)
    OxyPlot.SkiaSharp.Wpf.OxySKElement.OnRender(System.Windows.Media.DrawingContext)
    System.Windows.UIElement.Arrange(System.Windows.Rect)
    System.Windows.Controls.Grid.ArrangeOverride(System.Windows.Size)
    System.Windows.FrameworkElement.ArrangeCore(System.Windows.Rect)
    System.Windows.UIElement.Arrange(System.Windows.Rect)
    System.Windows.Controls.Border.ArrangeOverride(System.Windows.Size)
    System.Windows.FrameworkElement.ArrangeCore(System.Windows.Rect)
    System.Windows.UIElement.Arrange(System.Windows.Rect)
    ...
    [Call Stack Truncated]
```

### Steps to reproduce

1. Use a docking manager like syncfusion DockingManger
2. Undock the window with the plot, so that the window with the PlotView is no longer a descendant of the MainWindow in the VisualTree .
VisualTree of the Application with floating window:
image

Platform: x86 & x64
.NET version: 4.7.2
Latest OxyPlot Version (nuget packages build from develop brunch)

### Actual behaviour

The call Window.GetWindow(this) in OxySKElement.OnRender() returns always the MainWindow. I think it uses the logical tree not the visual tree and this.TransformToAncestor() fails, because the MainWindow is actually not an ancestor of the plot in case of a floating window.

### Possible Fix

Get the ancestor from the visual tree with a methode like this

```      
private Window GetAncestorWindowFromVisualTree(DependencyObject startElement)
        {
            DependencyObject parent = startElement;
            while (!(parent is Window))
            {
                if (parent == null) { break; }
                parent = VisualTreeHelper.GetParent(parent);
            }
            return parent as Window ?? Window.GetWindow(this);
        }
```

`var visualOffset = this.TransformToAncestor(GetAncestorWindowFromVisualTree(this)).Transform(default);`

### PR DESCRIPTION
**Pull request by albtaeler**
Fixes #1787.

### Checklist

- [ ] I have included examples or tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file
- [x] I have cleaned up the commit history (use rebase and squash)

### Changes proposed in this pull request:

- Get the window object that hosts the plot from the visual tree, not using Window.GetWindow()

My changes has a duplication of GetAncestorWindowFromVisualTree() in OxySKElement, but I think this is only a workaround, this should be fixed in the original SkiaSharp SKElement.

@oxyplot/admins

### CODE REVIEW

File: Source/OxyPlot.SkiaSharp.Wpf/OxySKElement.cs
Comment (VisualMelon) [inline Source/OxyPlot.SkiaSharp.Wpf/OxySKElement.cs:162]: Missing comments (per the `PlotView` version)

**Review by VisualMelon — COMMENTED** [review]
Thanks!

Just missing documentation on one of the methods.

I'll do a proper function check tonight/tomorrow when I have time, and then give it a couple of days before merging should the other maintainers spot something.

File: Source/OxyPlot.SkiaSharp.Wpf/OxySKElement.cs
Comment (albtaeler) [reply Source/OxyPlot.SkiaSharp.Wpf/OxySKElement.cs:162]: I've omitted the documentation in OxySkElement, because there's nothing documented in the whole file. But I can add it.

Comment (VisualMelon) [reply Source/OxyPlot.SkiaSharp.Wpf/OxySKElement.cs:162]: I hadn't realised; I suppose that's because it's a C&P of the provision by SkiaSharp. It would be nice to include it, given it is a divergence from the 'original' (indeed, we should consider documenting the whole file at some point).

**Review by VisualMelon — APPROVED** [review]

**Review by Jonarw — APPROVED** [review]
Thanks for spotting and fixing this issue!
I don't mind the lack of documentation in `OxySKElement` - as you said, this whole class is mostly a workaround, ideally this would be fixed on SkiaSharp side. (Or we may even be able to get rid of the `SKElement` completely in the future as it is just another container).

**Comment by VisualMelon** [comment]
I'm happy to merge without the comment; if #1792 goes anywhere, I'll add the documentation then (it makes changes to `OxySKElement`)

**Pull request merged**
